engineering: domain model in DDD (#42)
dwarvesf/brain: Engineering/Domain model in Domain Driven Design.md

---
tags: engineering, domain-driven-design, DDD
author: M.Vu Cuong(Jim)
---

> ## Domain model is

Reviewer: ```suggestion
## What is Domain Model
```

> ### Domain model is

Reviewer 2: https://github.com/dwarvesf/brain/blob/master/CONTRIBUTING.md
Last nit picks; use ## H2 headers as your top level header.

> Your organized and structured **knowledge of the business problem**. Represents as a diagram, code examples, or written documentation, and **must** be accessible and understandable by **everyone** involved with the project.

Reviewer: ```suggestion
In [Overview of Domain Driven Design | Domain Driven Design], Domain Model is the organized and structured **knowledge of the business problem**. Represents as a diagram, code examples, or written documentation, and **must** be accessible and understandable by **everyone** involved with the project.
```

> [Overview of Domain Driven Design](https://github.com/dwarvesf/brain/blob/master/Engineering/Overview%20of%20Domain%20Driven%20Design.md)

Reviewer: Remove this

> ## Domain model consists of

Reviewer: ```suggestion
## Components of Domain Model
```

> ### Domain model consists of

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

> The vocabulary and key concepts of the domain, and the relationships among all of the entities.
> These are mainly divided into: **Domain events**, **Commands**, **Aggregates**, and **Bounded context**.

Reviewer: ```suggestion
The vocabulary and key concepts of the domain, and the relationships among all of the entities. These are mainly divided into: **Domain events**, **Commands**, **Aggregates**, and **Bounded context**.
```

> ## Domain events

Reviewer: ```suggestion
### Domain events
```

> ### Domain events

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

*A statement in past tense* describes the **things that happened** in a business system that alternates the state of the entity.
Exp: Order submitted, Shopping cart updated

> ## Commands

Reviewer: ```suggestion
### Commands
```

> ### Commands

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

*A verb in present tense* describes the action that triggers the corresponding **domain event**. It is either user or system actions.
- Exp: `Add product`(Command) -> `Shopping cart updated`(Domain event)

> ## Aggregrates

Reviewer: ```suggestion
### Aggregrates
```

> ### Aggregrates

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

Represented by a *minimal cluster* of associated objects(domain events, commands, and actors) that we treat as a unit for data change. Each has a boundary and only exposes its root(**Aggregate Root**) which allows other objects to reference it.
- Exp: A team wants to update its member role according to each project: </n>
`Project, Update member role` (Aggregate Root) -> `Project's member role updated`</n>

> ## Bounded context

Reviewer: ```suggestion
### Bounded context
```

> ### Bounded context

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

A high-level structure consists of categorizations of functionality, represents a circle or square, that groups related entities together. It can bound parts of an aggregate or multiple aggregates.
- Exp: In an aggregate for the shopping process,  we draw the bounded contexts for **Shopping cart**, and **Offers**: </n>
Shopping cart(`User` -> `Add product to cart` -> `Cart updated`) -> Offers(`Promotiational Offers Identified` -> `Offers added`)

> ### References

Reviewer 2: Likewise, use ## H2 headers to mirror your top level header.

---
tags: engineering, domain-driven-design, DDD
author: M.Vu Cuong(Jim)
---

### Recap on Domain Driven Design

A software design approach to create a rich and meaningful **Domain model** which enforces the conversation around the domain. As result, removes _the communication lag_ between **Developers** and **Domain/Business experts**.

> [Here is the full overview](obsidian://open?vault=brain&file=Engineering%2FOverview%20of%20Domain%20Driven%20Design)

Reviewer: I think including the recap is kind of unnecessary. Maybe just mention the DDD article from the introduction of Domain model. e.g:

```
In [Domain Driven Design](link), Domain Model is ....
```

### Domain model is
Your organized and structured **knowledge of the business problem**. Represents as a diagram, code examples, or written documentation, and **must** be accessible and understandable by **everyone** involved with the project.
### Domain model consists of
The vocabulary and key concepts of the domain, and the relationships among all of the entities.
These are mainly divided into: **Domain events**, **Commands**, **Aggregates**, and **Bounded context**.
### Domain events
They are factual statements about the **things that happened** in a business system. It is important to phrase these statements in the past tense so one can frame this as a ‘what happened’ statement.
Exp: Order submitted, Shopping cart updated
### Commands
Is the identifier to why the **domain events** occurred. Commands express our intent for something to happen in the future as used in present tense. Commands may be documented as both user and system actions.
Exp: **Add product(Command)** -> Shopping cart updated(Domain event)
### Aggregrates

> It is represented by a cluster of events with corresponding commands and the responsible actor.

Reviewer 2: This statement is not unique and is taken directly from the reference source:
- https://creately.com/blog/diagrams/event-storming/

Please paraphrase this statement.
